File: javascript/dnd-character/dnd-character.js
const randomDice = () => Math.floor(Math.random() * 6) + 1;

export function abilityModifier(ability) {
    if (ability < 3) {
        throw new Error('Ability scores must be at least 3');
    } else if (ability > 18) {
        throw new Error('Ability scores can be at most 18');
    }
    return Math.floor((ability - 10) / 2);
}

export class Character {
    constructor() {
        this.strength = Character.rollAbility();
        this.dexterity = Character.rollAbility();
        this.constitution = Character.rollAbility();
        this.intelligence = Character.rollAbility();
        this.wisdom = Character.rollAbility();
        this.charisma = Character.rollAbility();
        this.hitpoints = 10 + abilityModifier(this.constitution);
    }

    static rollAbility() {
        return [randomDice(), randomDice(), randomDice(), randomDice()].sort((a, b) => b - a).slice(0, 3).reduce((prev, cur) => prev + cur);
    }
}

File: nim/dnd-character/dnd_character.nim
import std/[algorithm, random, sequtils]

type
  Character* = object
    strength*: int
    dexterity*: int
    constitution*: int
    intelligence*: int
    wisdom*: int
    charisma*: int
    hitpoints*: int

let baseHitPoints* = 10

proc ability*: int =
  var
    die = @[rand(1..6), rand(1..6), rand(1..6), rand(1..6)]
    sortedDie = die.sorted(Descending)
  sortedDie.delete(3)
  foldl(sortedDie, a + b)

proc modifier*(n: int): int =
  (n div 2) - 5

proc initCharacter*: Character =
  var
    c = Character(
      strength: ability(),
      dexterity: ability(),
      constitution: ability(),
      intelligence: ability(),
      wisdom: ability(),
      charisma: ability()
    )
  c.hitpoints = baseHitPoints + modifier(c.constitution)
  c

randomize()
